Add Envoy to proxy between client and server

sebastiantoh opened the pull request:
Implemented using a single envoy container
- Tried containerising the client, server as well as DB as well. But had difficulties configuring them. (one reason is due to the proto files being re-generated on the fly)
- `envoy.yaml` may not work on Windows / Mac hosts. May need to change lines 55 to either `host.docker.internal` or `docker.for.mac.localhost`

Some other things I want to clarify before I continue working on this:
- Any suggestions on how we should structure the GRPC clients in the frontend? Should we have a single client servicing all the requests, or create a client for each request?
- Should we directly use the proto's type definitions or map them to a separate type?


Tested using:
- Docker version 20.10.8
- docker-compose version 1.27.4
- Linux host

To test this PR, in separate shells, run `docker-compose up` in the root dir, `make run` in the `backend` dir, `npm run serve` in the `frontend` dir. 
- Add a `console.log(resp.getGemsList())` to lines 48 of `frontend/src/pages/Home/HomePage.vue` to see the gems being fetched (assuming the DB has been seeded with `make seed`)

sebastiantoh committed:
Add Envoy as single docker container (not working)

For some reason, it doesn't connect to the backend :'(

sebastiantoh committed:
Fix connection to server

sebastiantoh committed:
Remove commented out code

ianyong commented:
> Any suggestions on how we should structure the GRPC clients in the frontend? Should we have a single client servicing all the requests, or create a client for each request?

A single client seems to make sense, right? Do you know whether a ProtoBuf service is meant to be the entire API of an application? i.e., Instead of having a service for each functionality, we have a service for each service/application. Since we only have a monolith backend, we only have one ProtoBuf service.

ianyong commented:
> Should we directly use the proto's type definitions or map them to a separate type?

In the backend, the type definitions for the domain and the requests/responses are not coupled. For the frontend, I guess up to you?

okg00dbye commented:
> > Any suggestions on how we should structure the GRPC clients in the frontend? Should we have a single client servicing all the requests, or create a client for each request?
> 
> A single client seems to make sense, right? Do you know whether a ProtoBuf service is meant to be the entire API of an application? i.e., Instead of having a service for each functionality, we have a service for each service/application. Since we only have a monolith backend, we only have one ProtoBuf service.

I think we can just have one as well, maybe as a global variable like $router

> > Should we directly use the proto's type definitions or map them to a separate type?
> 
> In the backend, the type definitions for the domain and the requests/responses are not coupled. For the frontend, I guess up to you?

Think we can try using them directly, which way do you think would be easier?

okg00dbye commented:
Thanks so much for setting this up btw!

sebastiantoh commented:
> > Any suggestions on how we should structure the GRPC clients in the frontend? Should we have a single client servicing all the requests, or create a client for each request?
> 
> A single client seems to make sense, right? Do you know whether a ProtoBuf service is meant to be the entire API of an application? i.e., Instead of having a service for each functionality, we have a service for each service/application. Since we only have a monolith backend, we only have one ProtoBuf service.

I think it can be split. For the Envoy proxy, I think the definition of a "service" is different from the Protobuf "service".

sebastiantoh commented:
> > > Any suggestions on how we should structure the GRPC clients in the frontend? Should we have a single client servicing all the requests, or create a client for each request?
> > 
> > 
> > A single client seems to make sense, right? Do you know whether a ProtoBuf service is meant to be the entire API of an application? i.e., Instead of having a service for each functionality, we have a service for each service/application. Since we only have a monolith backend, we only have one ProtoBuf service.
> 
> I think we can just have one as well, maybe as a global variable like $router
> 
> > > Should we directly use the proto's type definitions or map them to a separate type?

Sounds good, I'll see how this can be done!

> > In the backend, the type definitions for the domain and the requests/responses are not coupled. For the frontend, I guess up to you?
> 
> Think we can try using them directly, which way do you think would be easier?

hmm, I'll try using them directly, I think we'll likely need to do some mapping eventually (e.g. converting between timestamps and Dayjs type), but this mapping should be quite minimal(?) for now

sebastiantoh committed:
Use env variables to retrieve server and port

sebastiantoh committed:
Remove committed binary

sebastiantoh committed:
Shift generated protobuf to within src

sebastiantoh committed:
Use Promise-based client

sebastiantoh committed:
Remove namespace for LatLng

sebastiantoh committed:
Reconfigure ESLint to fix false positives

sebastiantoh committed:
Add GemProvider

sebastiantoh committed:
Integrate gem map with backend

sebastiantoh committed:
Update User interface to include avatar

sebastiantoh committed:
Replace string with User in Gem interface

sebastiantoh committed:
Update README with Envoy instructions

sebastiantoh committed:
Add loading spinner

sebastiantoh committed:
Remove TODO

sebastiantoh commented:
> hmm, I'll try using them directly, I think we'll likely need to do some mapping eventually (e.g. converting between timestamps and Dayjs type), but this mapping should be quite minimal(?) for now

Decided against this. Was easier to map to a diff type haha

sebastiantoh commented:
Changes:
- Shift generated proto files to within `src`. Had difficulties accessing them otherwise
- Modified the settings of the proto generation. The new settings allows for promise-based clients as opposed to callback-based.
- Fixed some ESLint false positives (I hope?)
- Add GemProvider -- can access the methods in this provider by using `inject`
- The map should be integrated with the backend now

sebastiantoh commented:
Changes:
- Shift generated proto files to within `src`. Had difficulties accessing them otherwise
- Modified the settings of the proto generation. The new settings allows for promise-based clients as opposed to callback-based.
- Fixed some ESLint false positives (I hope?)
- Add GemProvider -- can access the methods in this provider by using `inject`
- Map is integrated with the backend now 

Will need to update your .env file in the frontend to include 2 new variables: 
```
VUE_APP_SERVER_URL=http://localhost
VUE_APP_SERVER_PORT=10000
```

sebastiantoh committed:
Fix typos

ianyong commented on envoy/envoy.yaml:
```diff
@@ -0,0 +1,56 @@
+admin:
+  access_log_path: /tmp/admin_access.log
+  address:
+    socket_address: { address: 0.0.0.0, port_value: 9901 }
```
Am I right to say that this is just for logging purposes? Is there any reason why the port is 9901?

sebastiantoh replied:
Hmm, it's not just logging. If you access localhost:9901, it also provides some sort of admin dashboard (though probably overkill for our usecase). 

I don't think there's any particular reason why it's 9901. But this seems to be the standard port based on the configs I've seen online

ianyong replied:
Ah I see, thanks!

ianyong commented on envoy/envoy.yaml:
```diff
@@ -0,0 +1,56 @@
+
+static_resources:
+  listeners:
+    - name: listener_0
```
Can this be more descriptive?

sebastiantoh replied:
Done! Hopefully it's more descriptive haha :sweat_smile:

ianyong commented on envoy/envoy.yaml:
```diff
@@ -0,0 +1,56 @@
+      http2_protocol_options: {}
+      lb_policy: round_robin
+      load_assignment:
+        cluster_name: cluster_0
```
Same for this.

ianyong reviewed (changes requested):
> Tried containerising the client, server as well as DB as well. But had difficulties configuring them. (one reason is due to the proto files being re-generated on the fly)

Am I right to say that they are currently not containerised? Seems a bit weird to have the reverse proxy be in a container but not the application itself.

ianyong commented on frontend/src/providers/Provider.vue:
```diff
@@ -0,0 +1,16 @@
+<template>
+  <GemProvider>
+    <slot></slot>
+  </GemProvider>
+</template>
```
To confirm, each ProtoBuf service has a corresponding `Provider` that is defined here, right?

sebastiantoh replied:
Doesn't necessarily have to be a strict 1-1 mapping between ProtoBuf service and the Provider. Although I guess most of the time it would be? But yep, any additional providers would have to be added here

ianyong replied:
Ok, thanks! I wasn't sure how having multiple ProtoBuf services would work since that would also indicate multiple clients.

sebastiantoh replied:
Hmm, you mean how it works with the proxy? 

My understanding is that if we initialise all the ProtoBuf services to send the request to the same listener (localhost:10000), it'll proxy all the requests to the same backend server. So in the big picture, there's still only 3 components: 1 client, 1 server and 1 proxy,

ianyong replied:
> My understanding is that if we initialise all the ProtoBuf services to send the request to the same listener (localhost:10000), it'll proxy all the requests to the same backend server. So in the big picture, there's still only 3 components: 1 client, 1 server and 1 proxy,

I see, thanks! Seemed a bit weird to me coz for REST APIs we typically have a single API abstraction that sends all requests. Whereas here, we have `GemServicePromiseClient`, `UserServicePromiseClient`, etc. I guess the `client` terminology threw me off.

sebastiantoh committed:
Use more descriptive names in Envoy config

sebastiantoh commented:
> Am I right to say that they are currently not containerised? Seems a bit weird to have the reverse proxy be in a container but not the application itself.

Yep, it's not atm. 

I have a simplified Docker image for both the frontend and backend, but it's not very nice haha. It involves manually `sed`-ing away the `preserve` hook. 
Also, I haven't set up the postgres image

ianyong commented:
> > Am I right to say that they are currently not containerised? Seems a bit weird to have the reverse proxy be in a container but not the application itself.
> 
> Yep, it's not atm.
> 
> I have a simplified Docker image for both the frontend and backend, but it's not very nice haha. It involves manually `sed`-ing away the `preserve` hook.
> Also, I haven't set up the postgres image

Okay, thanks! Guess those aren't very urgent anyway.

ianyong reviewed (approved):
LGTM!
